### PR TITLE
WIP towards commonmark

### PR DESCRIPTION
**Pull request by hayd**
cc @one-more-minute 

I've bundled these in together as they're quite a few varied areas when trying to fix commonmark spec. As [discussed elsewhere](https://github.com/one-more-minute/Markdown.jl/issues/7#issuecomment-80714218).

### CODE REVIEW

File: base/markdown/Common/block.jl [hunk 1]
```diff
@@ -146,12 +158,12 @@ function list(stream::IO, block::MD, config::Config)
...
             if fresh_line
-                skipwhitespace(stream)
-                if startswith(stream, b) != ""
+                sp = startswith(stream, r"^ {0,3}")
+                if !(startswith(stream, b) in [false, ""])
                     push!(the_list.items, parseinline(takebuf_string(buffer), config))
                     buffer = IOBuffer()
                 else
-                    write(buffer, ' ')
+                    write(buffer, '\n', sp)
```
Comment (hayd) [inline base/markdown/Common/block.jl]: not sure this is correct... but nested lists/etc aren't handled currently.

File: base/markdown/parse/util.jl [hunk 2]
```diff
@@ -46,7 +46,7 @@ function linecontains(io::IO, chars; allow_whitespace = true,
     l = readline(io) |> chomp
     length(l) == 0 && return allowempty
 
-    result = false
+    result = true
```
Comment (hayd) [inline base/markdown/parse/util.jl]: This should be `allowempty` rather than `true`.

**Comment by hayd** [comment]
@one-more-minute There's quite a few things checked off here (I can squash if necessary, but IMO these commits mostly make sense individually). There's some patterns to refactor in future, the main one:

    # three spaces is ok, four is a code block (shouldn't count according to the spec)
    startswith(stream, r"^ {0,3}")
    startswith(stream, " ") && return false
    ....

I think this also makes the ordering of the parsing functions less sensitive, but I'm not sure on the rules/spec when we have nesting...

---

One thing I see (and was true before) but can't explain is the following. This shouldn't be a hr, for the reason above (it's over 3 spaces indented), but it parses as an hr:
```
"""> Foo
                        ---"""|> Markdown.parse
```

**Comment by one-more-minute** [comment]
Ok great, looks good. It'll take me a little while to go over everything, and I'll do some tweaking/refactoring as well I think.

By the way, I just tried `Markdown.readme("Lazy")` and the new table code threw an error. Not related to these changes, I don't think, but worth looking into.

**Comment by shashi** [comment]
+1 This is great!

**Comment by hayd** [comment]
@one-more-minute pushed a oneline for fix for that.

Another README which doesn't work is Lexicon, perhaps as it overrides `@doc` ? (it does the same before this PR.)

```
help?> Lexicon
search: Lexicon

INFO: Parsing documentation for Lexicon.
INFO: Parsing documentation for Docile.
INFO: Parsing documentation for Docile.Interface.
```

**Comment by MichaelHatherly** [comment]
@hayd, that might be the use of [Link reference definitions](http://spec.commonmark.org/0.18/#link-reference-definition) in the README.

(Great work btw!)

**Comment by hayd** [comment]
@MichaelHatherly hmmm that's part of it (it won't parse as intended) - when parsing the README as a string I got an EOF error. Appended a one line fix. Will see if that'll fix the `?Lexicon`.

(@one-more-minute the OSX fail isn't me. Will try and refrain from adding more commits so you can review!!)

**Comment by hayd** [comment]
@one-more-minute thanks!

**Comment by one-more-minute** [comment]
No problem! This is great stuff.

By the way, I took out some of the list/hr changes for now. In general, parsers shouldn't have to know about each other, so that was a very suspicious special case. Happy to discuss other approaches to that though.

FYI, I plan to do some refactoring and wider changes tomorrow (hopefully the early merge save you some nasty rebasing ;)

**Comment by hayd** [comment]
@one-more-minute that's fair, I knew it was bad-form as I did it!

**Comment by hayd** [comment]
@MichaelHatherly that did fix it, though still has the INFO lines.
````
help?> Lexicon
search: Lexicon

 Lexicon
≡≡≡≡≡≡≡≡≡

  [![Build Status][travis-img]][travis-url] [![Build status][appveyor-img]][appveyor-url] [![Coverage Status][coveralls-img]][coveralls-url]
  [![Lexicon][pkgeval-img]][pkgeval-url]

  Lexicon.jl provides access to the documentation created by the @doc macro from [Docile.jl][docile-url]. It allows querying of package documentation from
  the Julia REPL and building HTML documentation.

 Installation
==============

  Lexicon.jl is available from METADATA and may be installed via:

  Pkg.add("Lexicon")

 About
=======

  Docile.jl retains the documentation layer (@doc, @docstrings, @md_mstr), while Lexicon.jl takes over the presentation of documentation stored by Docile.jl.

  Lexicon.jl separates the non-essential parts from Docile.jl so that package load times are not impacted when documenting packages using @doc. See [this
  issue][issue-url] for details regarding the change.

 Documentation
===============

  Lexicon.jl-generated documentation is available [here][docs-url].

  [travis-img]: https://travis-ci.org/MichaelHatherly/Lexicon.jl.svg?branch=master [travis-url]: https://travis-ci.org/MichaelHatherly/Lexicon.jl

  [appveyor-img]: https://ci.appveyor.com/api/projects/status/qmuv67ku625ioiwc/branch/master [appveyor-url]:
  https://ci.appveyor.com/project/MichaelHatherly/lexicon-jl/branch/master

  [coveralls-img]: https://img.shields.io/coveralls/MichaelHatherly/Lexicon.jl.svg [coveralls-url]: https://coveralls.io/r/MichaelHatherly/Lexicon.jl

  [pkgeval-img]: http://pkg.julialang.org/badges/Lexicon_release.svg [pkgeval-url]: http://pkg.julialang.org/?pkg=Lexicon&ver=release

  [docile-url]: https://github.com/MichaelHatherly/Docile.jl

  [issue-url]: https://github.com/MichaelHatherly/Docile.jl/issues/27

  [docs-url]: https://michaelhatherly.github.io/Lexicon.jl/index.html
INFO: Parsing documentation for Docile.Interface.
INFO: Parsing documentation for Lexicon.
INFO: Parsing documentation for Docile.
````

**Comment by MichaelHatherly** [comment]
The INFO lines are coming from Lexicon searching through documentation and having to parse the source files. They only appear once per session, but could be silenced if people feel they're unnecessary.

**Pull request merged**